### ISSUE TITLE
haskoin-wallet: Support for arbitrary watch-only addresses

### ISSUE THREAD

**Issue by runeksvendsen** [jprupp/haskoin-core#290]
Hi there

I'm planning on using haskoin-wallet for a backend service in a Bitcoin app, and in this particular app, the relevant Bitcoin addresses (which I would like to know the balances of) are not derived from the server's private key alone, but rather from a [`CHECKLOCKTIMEVERIFY`-style redeemScript](https://github.com/bitcoin/bips/blob/master/bip-0065.mediawiki#Twofactor_wallets) which contains the server's public key, plus a client public key and a timestamp.

So in other words, the Bitcoin address that is derived from the server's private key is never used for anything by itself, because the corresponding public key is only used in redeemScripts which also include a client pubkey and a timestamp.

Is there a way for me to get haskoin-wallet to retrieve information about arbitrary (P2SH) addresses, that I have derived from custom redeemScripts, which my app uses?

**Comment by jprupp** [comment]
Hello Rune,

There is no way to do this at the moment. Only regular and P2SH multi-signature addresses are currently supported. Patch would be nice.

Good luck!

**Comment by runeksvendsen** [comment]
Hello

Alright, I'm strongly considering going ahead and implementing this.

I'm thinking the way to do it is to introduce a new "watch-only" account type, and also a new `addaddr` command, which allows adding arbitrary addresses to this watch-only account.

How would you recommend implementing this? As far as I can see, the only thing that would need to change for the `Account` type is that the `accountKeys` field needs to be able to contain addresses too, rather than only PubKeys. It would work this way, but it may be trying to squeeze this watch-only concept into where it doesn't fit. As far as I can see, for the watch-only account type, we really only need the fields `accountName`, `accountAddreses` (which would contain the watch-only addresses) and `accountCreated`; all other fields wouldn't apply.

Let me know what you think.

**Comment by runeksvendsen** [comment]
Oh, wait, is `accountKeys` for multi-sig pubkeys? In that case, putting watch-only addresses in there doesn't seem appropriate at all. Anyway, you probably have a better overview of the code base, so I will await your advice.

**Comment by runeksvendsen** [comment]
We could also add an `accountWatchAddr :: [Address]` field to the `Account` type, such that we don't introduce a new account type, but just give watch-only functionality to the existing account types.

That way, we wouldn't be changing the existing logic for accounts. So, when constructing the Bloom filter for the node, we just add the regular wallet addresses as we always do, and then add the watch-only addresses from `accountWatchAddr` (if there are any).

**Comment by jprupp** [comment]
I would have to read the `haskoin-wallet` code to acquaint myself with it again in order to be able to give you any specific advise. I will try to get you as much help as I can without doing that for now.

There are some assumptions that were made when the wallet was written. Some of those assumptions are that addresses are always generated from one or more extended public keys following the tree format established by [BIP-32](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki). If these assumptions no longer hold for your proposed address generation scheme, then changes are going to have to be made in lots of places. I will let you figure out the best approach to proceed, since you are better aware of the model that you want to use to generate new addresses in these accounts that you need to create. I hope that the code is easy to change. I did a massive re-engineering of this codebase before, and it was not particularly difficult.

I can give you a bit of an overall design philosophy information:
- `haskoin-core` contains mostly pure functions and highly-reusable code that is not wallet-specific.
- `haskoin-node` deals with the Bitcoin protocol and network. It synchronizes the header tree, relays transactions and block headers to the wallet, holds and synchronizes bloom filters among peers.
- `haskoin-wallet` is probably where all the changes that you need are to be done. This is the code where actual tracking of transactions and accounts is done. It makes heavy use of [esqueleto](https://www.stackage.org/lts-6.20/package/esqueleto-2.4.3) database queries. You will need to learn about that.

**Comment by runeksvendsen** [comment]
Alright, so I've had the chance to look more at the code, and, for now at least, I've decided against implementing this. It looks, to me at least, to be a relatively complex change, so I've settled for querying the P2SH addresses in a different way.

The whole issue of **1)** how to specify the parameters (pubkeys, CLTV timestamps etc.) required to construct P2SH redeemScripts and **2)** how to construct the (possibly multiple different) scriptSigs that can redeem a given redeemScript is fairly complex. I think it would be a much better fit in a library of its own, which would implement a Bitcoin script-Embedded Domain-Specific Language. Someone, with more experience in implementing EDSLs, could make something really awesome by creating a type-safe Bitcoin script EDSL, using the types from `haskoin-core`. Like `esqueleto` but for Bitcoin script rather than SQL.

**Comment by jprupp** [comment]
That EDSL system for Bitcoin seems like an interesting proposition. I'd love to see it happen.

**Comment by runeksvendsen** [comment]
Indeed! I'm afraid I don't have the skill, Haskell-wise, to make it happen though. At least not at this time. But it would be pretty cool.

**Comment by runeksvendsen** [comment]
I changed my mind. Watch-only functionality is too awesome not to have, so I've started implementing it. It's working now:

```
rune@Runes-MacBook-Pro:~/IdeaProjects/haskoin-fork$ hw newwatch watch
Account : watch
rune@Runes-MacBook-Pro:~/IdeaProjects/haskoin-fork$ hw watchadd watch 121CEyDDEewZfp7mpFLd6WvfD9fcyWpwQe
1 : 121CEyDDEewZfp7mpFLd6WvfD9fcyWpwQe
rune@Runes-MacBook-Pro:~/IdeaProjects/haskoin-fork$ hw monitor watch
Id         : 8728a903567914c1c2ce2e493df5fbc53b8556c0750c5a0126fb13321280db0c
Address    : Address "121CEyDDEewZfp7mpFLd6WvfD9fcyWpwQe" (watch-only)
Tx         : Tx "0100000001d78c08960c4c8177913ed27c69c70401c960ed7964b15c0165a9971b0cf44de0010000006b483045022100f36c1d54002bbb2f18753df643af6cf341bf221e6379d326ddaec59085167aa90220239c6cd703dad9f521c3a7b2bd0602718dc6901468d0e3f8d9fdf913f627a13c01210297918cf1417b724efab9ced492bf81727569208b23b6dfbf8cf5b9d341553918ffffffff0248460300000000001976a9140b018259015c6409f7b73488da79ae027190b16188acbc503801000000001976a914f5a41da804708b8a440c66fb959b53d1f8bb47f188ac00000000"

Id         : 8728a903567914c1c2ce2e493df5fbc53b8556c0750c5a0126fb13321280db0c
Address    : Address "121CEyDDEewZfp7mpFLd6WvfD9fcyWpwQe" (watch-only)
Tx         : Tx "0100000001d78c08960c4c8177913ed27c69c70401c960ed7964b15c0165a9971b0cf44de0010000006b483045022100f36c1d54002bbb2f18753df643af6cf341bf221e6379d326ddaec59085167aa90220239c6cd703dad9f521c3a7b2bd0602718dc6901468d0e3f8d9fdf913f627a13c01210297918cf1417b724efab9ced492bf81727569208b23b6dfbf8cf5b9d341553918ffffffff0248460300000000001976a9140b018259015c6409f7b73488da79ae027190b16188acbc503801000000001976a914f5a41da804708b8a440c66fb959b53d1f8bb47f188ac00000000"

Id         : 8728a903567914c1c2ce2e493df5fbc53b8556c0750c5a0126fb13321280db0c
Address    : Address "121CEyDDEewZfp7mpFLd6WvfD9fcyWpwQe" (watch-only)
Tx         : Tx "0100000001d78c08960c4c8177913ed27c69c70401c960ed7964b15c0165a9971b0cf44de0010000006b483045022100f36c1d54002bbb2f18753df643af6cf341bf221e6379d326ddaec59085167aa90220239c6cd703dad9f521c3a7b2bd0602718dc6901468d0e3f8d9fdf913f627a13c01210297918cf1417b724efab9ced492bf81727569208b23b6dfbf8cf5b9d341553918ffffffff0248460300000000001976a9140b018259015c6409f7b73488da79ae027190b16188acbc503801000000001976a914f5a41da804708b8a440c66fb959b53d1f8bb47f188ac00000000"
```

But it still needs more work. The change is quite a lot bigger than my previous commits, although it mostly constitutes separating out the Bloom stuff from `Network.Haskoing.Wallet.Accounts` and creating a new account type -- `WatchAccount` -- in which I put a new type called `WatchAddr`. So the watch-only stuff doesn't mix with the regular wallet stuff in the database, which makes the change a lot cleaner, and without as many code changes.

**Comment by jprupp** [comment]
That's great!

I am currently too busy with segwit-related stuff at work to help you out at the moment. Perhaps after the rollout of segwit in production I can be of more assistance. I hope you nail it. It is fun to refactor code in Haskell, mainly because it is very easy.

Cheers!

**Comment by ysangkok** [comment]
@runeksvendsen what happened? Why not merge it?

**Comment by runeksvendsen** [comment]
@ysangkok The code never reached a quality I was happy with. In the end I realized that the right way to do it would be to abstract over the `Account` type with a type class, and define wallet operations on that type class, such that code can be shared with other account types. I ended up having two different account types (the standard `Account` and the watch-only account), and most of the operations were duplicates of each other, because they shared a lot of features. So the code that came out of it was mostly unusable, but the solution is to abstract over the `Account` type to allow different implementations, such as watch-only.
